File: packages/risk-model/src/utils.test.ts
import {
  parseDuration,
  getPositionKey,
  calculatePnlPct,
  calculateMarginUsagePct,
  formatPct,
  deepMerge,
} from './utils.js';

describe('parseDuration', () => {
  it('should return number as-is', () => {
    expect(parseDuration(1000)).toBe(1000);
    expect(parseDuration(0)).toBe(0);
  });

  it('should parse milliseconds', () => {
    expect(parseDuration('100ms')).toBe(100);
  });

  it('should parse seconds', () => {
    expect(parseDuration('10s')).toBe(10000);
  });

  it('should parse minutes', () => {
    expect(parseDuration('15m')).toBe(900000);
    expect(parseDuration('1m')).toBe(60000);
  });

  it('should parse hours', () => {
    expect(parseDuration('1h')).toBe(3600000);
    expect(parseDuration('2h')).toBe(7200000);
  });

  it('should parse days', () => {
    expect(parseDuration('1d')).toBe(86400000);
  });

  it('should parse decimal values', () => {
    expect(parseDuration('1.5h')).toBe(5400000);
  });

  it('should throw on invalid format', () => {
    expect(() => parseDuration('invalid')).toThrow('Invalid duration format');
    expect(() => parseDuration('10x')).toThrow('Invalid duration format');
  });
});

describe('getPositionKey', () => {
  it('should generate correct key', () => {
    expect(getPositionKey('BTCUSDT', 'LONG')).toBe('BTCUSDT:LONG');
    expect(getPositionKey('ETHUSDT', 'SHORT')).toBe('ETHUSDT:SHORT');
  });
});

describe('calculatePnlPct', () => {
  it('should calculate PnL percentage correctly', () => {
    expect(calculatePnlPct({ unrealizedPnl: 100, marginUsed: 1000 })).toBe(0.1);
    expect(calculatePnlPct({ unrealizedPnl: -50, marginUsed: 100 })).toBe(-0.5);
    expect(calculatePnlPct({ unrealizedPnl: 200, marginUsed: 100 })).toBe(2.0);
  });

  it('should return 0 when marginUsed is 0', () => {
    expect(calculatePnlPct({ unrealizedPnl: 100, marginUsed: 0 })).toBe(0);
  });
});

describe('calculateMarginUsagePct', () => {
  it('should calculate margin usage correctly', () => {
    expect(calculateMarginUsagePct({ marginUsed: 5000, equity: 10000 })).toBe(0.5);
    expect(calculateMarginUsagePct({ marginUsed: 8000, equity: 10000 })).toBe(0.8);
  });

  it('should return 0 when equity is 0', () => {
    expect(calculateMarginUsagePct({ marginUsed: 100, equity: 0 })).toBe(0);
  });
});

describe('formatPct', () => {
  it('should format percentage correctly', () => {
    expect(formatPct(0.5)).toBe('50.00%');
    expect(formatPct(0.1234)).toBe('12.34%');
    expect(formatPct(1.5)).toBe('150.00%');
  });

  it('should respect decimal places', () => {
    expect(formatPct(0.12345, 1)).toBe('12.3%');
    expect(formatPct(0.12345, 4)).toBe('12.3450%');
  });
});

describe('deepMerge', () => {
  it('should merge simple objects', () => {
    const target = { a: 1, b: 2 };
    const source = { b: 3 };
    expect(deepMerge(target, source)).toEqual({ a: 1, b: 3 });
  });

  it('should merge nested objects', () => {
    const target = { a: { x: 1, y: 2 }, b: 3 };
    const source: Partial<typeof target> = { a: { x: 1, y: 10 } };
    expect(deepMerge(target, source)).toEqual({ a: { x: 1, y: 10 }, b: 3 });
  });

  it('should not mutate original objects', () => {
    const target = { a: 1, b: 2 };
    const source = { a: 10 };
    const result = deepMerge(target, source);
    expect(target).toEqual({ a: 1, b: 2 });
    expect(source).toEqual({ a: 10 });
    expect(result).toEqual({ a: 10, b: 2 });
  });

  it('should handle undefined values in source', () => {
    const target = { a: 1, b: 2 };
    const source: Partial<typeof target> = { a: undefined };
    expect(deepMerge(target, source)).toEqual({ a: 1, b: 2 });
  });
});

File: pkg/risk/utils_test.go
package risk

import (
	"testing"
	"time"
)

func TestParseDuration(t *testing.T) {
	tests := []struct {
		name     string
		input    interface{}
		expected int64
		wantErr  bool
	}{
		{"milliseconds", 1000, 1000, false},
		{"15 minutes", "15m", 900000, false},
		{"1 hour", "1h", 3600000, false},
		{"2 days", "2d", 172800000, false},
		{"30 seconds", "30s", 30000, false},
		{"500 milliseconds string", "500ms", 500, false},
		{"invalid format", "15x", 0, true},
		{"invalid type", true, 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := parseDuration(tt.input)
			if tt.wantErr {
				if err == nil {
					t.Errorf("Expected error but got none")
				}
				return
			}
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
				return
			}
			if result != tt.expected {
				t.Errorf("Expected %d, got %d", tt.expected, result)
			}
		})
	}
}

func TestGetPositionKey(t *testing.T) {
	tests := []struct {
		symbol   string
		side     PositionSide
		expected string
	}{
		{"BTCUSDT", PositionSideLong, "BTCUSDT:LONG"},
		{"ETHUSDT", PositionSideShort, "ETHUSDT:SHORT"},
	}

	for _, tt := range tests {
		t.Run(tt.expected, func(t *testing.T) {
			result := GetPositionKey(tt.symbol, tt.side)
			if result != tt.expected {
				t.Errorf("Expected %s, got %s", tt.expected, result)
			}
		})
	}
}

func TestCalculatePnlPct(t *testing.T) {
	tests := []struct {
		name     string
		position PositionSnapshot
		expected float64
	}{
		{
			name: "positive PnL",
			position: PositionSnapshot{
				MarginUsed:    1000,
				UnrealizedPnl: 500,
			},
			expected: 0.5,
		},
		{
			name: "negative PnL",
			position: PositionSnapshot{
				MarginUsed:    1000,
				UnrealizedPnl: -200,
			},
			expected: -0.2,
		},
		{
			name: "zero margin",
			position: PositionSnapshot{
				MarginUsed:    0,
				UnrealizedPnl: 100,
			},
			expected: 0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := CalculatePnlPct(tt.position)
			if result != tt.expected {
				t.Errorf("Expected %f, got %f", tt.expected, result)
			}
		})
	}
}

func TestCalculateMarginUsagePct(t *testing.T) {
	tests := []struct {
		name     string
		account  AccountSnapshot
		expected float64
	}{
		{
			name: "50% margin usage",
			account: AccountSnapshot{
				Equity:     10000,
				MarginUsed: 5000,
			},
			expected: 0.5,
		},
		{
			name: "80% margin usage",
			account: AccountSnapshot{
				Equity:     10000,
				MarginUsed: 8000,
			},
			expected: 0.8,
		},
		{
			name: "zero equity",
			account: AccountSnapshot{
				Equity:     0,
				MarginUsed: 1000,
			},
			expected: 0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := CalculateMarginUsagePct(tt.account)
			if result != tt.expected {
				t.Errorf("Expected %f, got %f", tt.expected, result)
			}
		})
	}
}

func TestFormatPct(t *testing.T) {
	tests := []struct {
		name     string
		value    float64
		decimals int
		expected string
	}{
		{"50%", 0.5, 2, "50.00%"},
		{"80.5%", 0.805, 2, "80.50%"},
		{"100%", 1.0, 2, "100.00%"},
		{"12.345%", 0.12345, 3, "12.345%"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := FormatPct(tt.value, tt.decimals)
			if result != tt.expected {
				t.Errorf("Expected %s, got %s", tt.expected, result)
			}
		})
	}
}

func TestDeepMerge(t *testing.T) {
	target := PositionRiskConfig{
		StopProfitPct: 1.5,
		StopLossPct:   0.8,
		OnBreach:      PositionBreachClosePosition,
		Cooldown:      "15m",
	}

	override := PositionRiskConfig{
		StopProfitPct: 2.0,
		StopLossPct:   1.0,
	}

	result := DeepMerge(target, override)

	if result.StopProfitPct != 2.0 {
		t.Errorf("Expected StopProfitPct 2.0, got %f", result.StopProfitPct)
	}
	if result.StopLossPct != 1.0 {
		t.Errorf("Expected StopLossPct 1.0, got %f", result.StopLossPct)
	}
	if result.OnBreach != PositionBreachClosePosition {
		t.Errorf("Expected OnBreach to remain as CLOSE_POSITION")
	}
}

func TestParseCooldown(t *testing.T) {
	tests := []struct {
		name     string
		input    interface{}
		expected time.Duration
		wantErr  bool
	}{
		{"15 minutes", "15m", 15 * time.Minute, false},
		{"1 hour", "1h", 1 * time.Hour, false},
		{"milliseconds", 60000, 60 * time.Second, false},
		{"invalid", "invalid", 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := ParseCooldown(tt.input)
			if tt.wantErr {
				if err == nil {
					t.Errorf("Expected error but got none")
				}
				return
			}
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
				return
			}
			if result != tt.expected {
				t.Errorf("Expected %v, got %v", tt.expected, result)
			}
		})
	}
}

func TestCreateInitialState(t *testing.T) {
	state := CreateInitialState()

	if state.AccountBlocked {
		t.Errorf("Expected AccountBlocked to be false")
	}
	if state.Positions == nil {
		t.Errorf("Expected Positions map to be initialized")
	}
	if len(state.Positions) != 0 {
		t.Errorf("Expected Positions map to be empty")
	}
}
